### PR TITLE
Add experimental addon flag to submission, edit listing;

### LINKED ISSUE #3084: Implement experimental flag for listed add-ons
This is part of #3083. Listed add-on developers need a way to flag their add-ons as experimental, so they can be listed but with reduced visibility.

* On Step 3 of the submission process, add a checkbox with the label "This add-on is experimental", not checked by default. I think this should show up between the summary and the categories. It should have a help tooltip explaining that experimental add-ons are listed on the site but have reduced visibility, and that this setting can be changed later.
* In the Developer Hub, under Edit Listing > Basic Information, add a field to edit this flag, also between the summary and categories.
* The experimental flag should be handled like Preliminary Review when it comes to displaying the install buttons for all versions of the add-on (yellow stripes), showing up after other add-ons in search results, and not appearing in the Disco Pane. Other limitations for preliminary add-ons, like contributions and automatic updates, don't apply to add-ons with this flag.
* The text shown with the yellow striped button should be changed from "This add-on has been preliminarily reviewed" to "This add-on has been flagged as experimental by its developers". The Learn More link should not be there.

Reviewers won't manipulate this flag, but can recommend developers to use it when relevant.

### PR DESCRIPTION
**Pull request by eviljeff**
show different install button; degrade search ranking.  Fixes #3084

### CODE REVIEW

**Comment by eviljeff** [comment]
@jvillalobos is the copy okay?  I can add screenshots if that helps.

File: src/olympia/addons/templates/addons/button.html [hunk 1]
```diff
@@ -98,6 +98,10 @@
       This add-on has been preliminarily reviewed by Mozilla.
       <a href="{{ url }}">Learn more</a>
     {% endtrans %} </p>
+  {% elif b.experimental %}
+    <p class="warning">{% trans url=url('pages.faq') + "#preliminary" %}
+      This add-on has been flagged as experimental by its developers.
```
Comment (jvillalobos) [inline src/olympia/addons/templates/addons/button.html]: "Flagged" may have negative connotations. "Marked"?

File: src/olympia/addons/templates/addons/impala/button.html [hunk 2]
```diff
@@ -61,6 +61,10 @@
       This add-on has been preliminarily reviewed by Mozilla.
       <a href="{{ url }}">Learn more</a>
     {% endtrans %} </p>
+  {% elif b.experimental %}
+    <p class="warning">{% trans url=url('pages.faq') + "#preliminary" %}
+      This add-on has been flagged as experimental by its developers
```
Comment (jvillalobos) [inline src/olympia/addons/templates/addons/impala/button.html]: Same as previous note.

File: src/olympia/devhub/templates/devhub/addons/edit/basic.html [hunk 3]
```diff
@@ -67,6 +67,21 @@
               {% endif %}
             </td>
           </tr>
+          {% if waffle.flag('no-prelim-review') %}
+          <tr>
+            <th>
+              {{ tip(_("Experimental?"),
+                     _("If your add-on on is experimental or otherwise not "
+                       "ready for a general use.  The add-on will be listed "
```
Comment (jvillalobos) [inline src/olympia/devhub/templates/devhub/addons/edit/basic.html]: Remove "a" before "general use".

File: src/olympia/devhub/templates/devhub/addons/edit/basic.html [hunk 4]
```diff
@@ -67,6 +67,21 @@
...
+                       "but will have reduced visibility. ")) }}
+            </th>
+            <td>
+              {{ flags(_("This add-on is experimental."),
+                       addon.is_experimental if not editable else form.is_experimental, editable,
+                       _("This add-on is not experimental, it's useful for general use.")) }}
```
Comment (jvillalobos) [inline src/olympia/devhub/templates/devhub/addons/edit/basic.html]: I'd just go with "This add-on is ready for general use."

File: src/olympia/devhub/templates/devhub/addons/submit/describe.html [hunk 5]
```diff
@@ -46,6 +46,20 @@
       </div>
     </div>
     {% if addon.is_listed %}
+      {% if waffle.flag('no-prelim-review') %}
+      <div class="addon-submission-field">
+        <label for="{{ form.is_experimental.auto_id }}">
+        {{ form.is_experimental }}
+        {{ _('This add-on is experimental') }}
+        <span class="tip tooltip"
+              title="{{ _('Check this option if your add-on on is experimental '
```
Comment (jvillalobos) [inline src/olympia/devhub/templates/devhub/addons/submit/describe.html]: Remove "on" after "add-on".

File: src/olympia/devhub/templates/devhub/addons/submit/describe.html [hunk 6]
```diff
@@ -46,6 +46,20 @@
...
+                          'or otherwise not ready for a general use.  The '
```
Comment (jvillalobos) [inline src/olympia/devhub/templates/devhub/addons/submit/describe.html]: Remove "a" before "general use".

**Comment by eviljeff** [comment]
@diox can you especially check I added the property to ES and the API response correctly - my ES locally was playing up so it was difficult to check everything was working as expected.

File: src/olympia/addons/indexers.py [hunk 7]
```diff
@@ -77,6 +77,7 @@ def get_mapping(cls):
                     'hotness': {'type': 'double'},
                     'icon_type': {'type': 'string', 'index': 'no'},
                     'is_disabled': {'type': 'boolean'},
+                    'is_experimental': {'type': 'boolean'},
```
Comment (diox) [inline src/olympia/addons/indexers.py]: I assume we don't need to search on it, so you can add `'index': 'no'`

Comment (eviljeff) [inline src/olympia/addons/indexers.py]: We could feasibly need to search on it - it's a psuedo-status - so I guess it should be treated how `status` is.

File: src/olympia/addons/forms.py [hunk 8]
```diff
@@ -156,10 +156,11 @@ class AddonFormBasic(AddonFormBase):
     summary = TransField(widget=TransTextarea(attrs={'rows': 4}),
                          max_length=250)
     tags = forms.CharField(required=False)
+    is_experimental = forms.BooleanField(initial=False, required=False)
```
Comment (diox) [inline src/olympia/addons/forms.py]: The `initial=False` here is suspicious. Shouldn't it follow the model instance value instead, whatever that currently is ?

File: src/olympia/addons/tests/test_serializers.py [hunk 9]
```diff
@@ -171,6 +171,12 @@ def test_is_source_public(self):
 
         assert result['is_source_public'] is True
 
+    def test_is_experimental(self):
+        self.addon = addon_factory(is_experimental=True)
+        result = self.serialize()
+
+        assert result['is_experimental'] is True
```
Comment (diox) [inline src/olympia/addons/tests/test_serializers.py]: Also modify `test_basic`, adding an `assert result['is_experimental'] is False`

File: src/olympia/legacy_api/tests/test_views.py [hunk 10]
```diff
@@ -1007,6 +1007,18 @@ def test_unlisted_are_ignored(self):
         self.assertContains(response, """<searchresults total_results="0">""")
         self.assertContains(response, "</addon>", 0)
 
+    def test_experimental_are_ignored(self):
+        """
+        Test that unlisted add-ons are not shown.
```
Comment (diox) [inline src/olympia/legacy_api/tests/test_views.py]: Wrong comment

File: src/olympia/addons/indexers.py (hunk 7 above)
Comment (diox) [inline src/olympia/addons/indexers.py]: Yeah and actually you're searching on it - in the legacy api search view. Carry on.

File: src/olympia/addons/forms.py (hunk 8 above)
Comment (eviljeff) [inline src/olympia/addons/forms.py]: copy/paste fail

File: src/olympia/addons/tests/test_serializers.py (hunk 9 above)
Comment (eviljeff) [inline src/olympia/addons/tests/test_serializers.py]: done

**Comment by diox** [comment]
You're missing a test that proves the devhub shows the field (with the correct value) when the waffle is on

r+with that

**Pull request merged**